File: java/lesson 01/TapeEquilibrium.java
package com.mangoturtle.codility;

/**
 * Created by Dean on 6/24/2015.
 *
 * TapeEquilibrium - Time Complexity
 * https://codility.com/demo/take-sample-test/tape_equilibrium/
 *
 * Detected time complexity: O(N)
 */
public class TapeEquilibrium {

    public int solution(int[] A) {
        int left = 0, right, result = Integer.MAX_VALUE;
        int sum = getSum(A);
        right = sum;

        for (int i = 0; i < A.length - 1; i++){
            left += A[i];
            right -= A[i];
            result = Math.min(result, Math.abs(left - right));
        }
        return result;
    }

    public int getSum(int[] A){
        int result = 0;
        for (int a: A)
            result += a;
        return result;
    }

    public static void main(String[] args){
        TapeEquilibrium tE = new TapeEquilibrium();
        System.out.println(tE.solution(new int[]{3, 1, 2, 4, 3}));
    }
}

File: objective-c/lesson01/TapeEquilibrium.m
#import <Foundation/Foundation.h>
//
// TapeEquilibrium - Time Complexity - Objective-C
// Minimize the value |(A[0] + ... + A[P-1]) - (A[P] + ... + A[N-1])|.
// https://codility.com/demo/take-sample-test/tape_equilibrium/
//
int solution(NSMutableArray *A);
int getSum(NSMutableArray *A);

int solution(NSMutableArray *A) {
	int right = getSum(A), left = 0, result = INT_MAX;
    for (int i = 0; i < [A count] - 1; i++){
        NSInteger item = [[A objectAtIndex: i] integerValue];
        left += item;
        right -= item;
        result = MIN(result, ABS(left - right));
    }
    return result;
}

int getSum(NSMutableArray *A){
    int result = 0;
    for (id a in A){
        result += [a integerValue];
    }
    return result;
}

int main(int argc, const char * argv[]) {
	@autoreleasepool {
		NSLog(@"result = %d", solution([NSMutableArray arrayWithObjects: @3, @1, @2, @4, @3, nil]));
	}
	return 0;
}